domain/removal/state/application.go

```diff
@@ -0,0 +1,300 @@
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	applicationUUIDCount := entityAssociationCount{UUID: aUUID}
+
+	stmt, err := st.Prepare(`
+SELECT COUNT(*) AS &entityAssociationCount.count
+FROM   application
+WHERE  uuid = $entityAssociationCount.uuid;`, applicationUUIDCount)
+	if err != nil {
+		return errors.Errorf("preparing application life query: %w", err)
+	}
+
+	deleteApplicationStmt, err := st.Prepare(`
+DELETE FROM application
+WHERE  uuid = $entityAssociationCount.uuid;`, applicationUUIDCount)
+	if err != nil {
+		return errors.Errorf("preparing unit delete: %w", err)
+	}
+
+	return errors.Capture(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx, stmt, applicationUUIDCount).Get(&applicationUUIDCount)
+		if errors.Is(err, sqlair.ErrNoRows) || applicationUUIDCount.Count == 0 {
+			return applicationerrors.ApplicationNotFound
+		} else if err != nil {
+			return errors.Errorf("running application life query: %w", err)
+		}
+
+		if err := st.deleteApplicationAnnotations(ctx, tx, aUUID); err != nil {
+			return errors.Errorf("deleting application annotations: %w", err)
+		}
+
+		if err := st.deleteCloudServices(ctx, tx, aUUID); err != nil {
+			return errors.Errorf("deleting cloud services: %w", err)
+		}
+
+		if err := st.deleteDeviceConstraintAttributes(ctx, tx, aUUID); err != nil {
+			return errors.Errorf("deleting device constraint attributes: %w", err)
+		}
+
+		if err := st.deleteSimpleApplicationReferences(ctx, tx, aUUID); err != nil {
+			return errors.Errorf("deleting simple application references: %w", err)
+		}
+
+		if err := tx.Query(ctx, deleteApplicationStmt, applicationUUIDCount).Run(); err != nil {
+			return errors.Errorf("deleting application: %w", err)
+		}
+
+		return nil
+	}))
+}
+
+func (st *State) deleteSimpleApplicationReferences(ctx context.Context, tx *sqlair.TX, aUUID string) error {
+	app := applicationUUID{UUID: aUUID}
+
+	for _, table := range []string{
+		"application_channel",
+		"application_platform",
+		"application_scale",
+		"application_config",
+		"application_config_hash",
+		"application_constraint",
+		"application_setting",
+		"application_exposed_endpoint_space",
+		"application_exposed_endpoint_cidr",
+		"application_endpoint",
+		"application_extra_endpoint",
+		"application_storage_directive",
+		"application_workload_version",
+		"device_constraint",
+	} {
+		deleteApplicationReference := fmt.Sprintf(`DELETE FROM %s WHERE application_uuid = $applicationUUID.application_uuid`, table)
+		deleteApplicationReferenceStmt, err := st.Prepare(deleteApplicationReference, app)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		if err := tx.Query(ctx, deleteApplicationReferenceStmt, app).Run(); err != nil {
+			return errors.Errorf("deleting reference to application in %s: %w", table, err)
+		}
+	}
+	return nil
+}
+
+func (st *State) deleteCloudServices(ctx context.Context, tx *sqlair.TX, aUUID string) error {
```

Comment:
I think the link layer device and the ip address(es) should be removed as well, right?

Reply:
Probably, but that can be done with ll epic